rust/cymbal/src/lib.rs

```diff
@@ -7,3 +18,130 @@ pub mod langs;
 pub mod metric_consts;
 pub mod symbol_store;
 pub mod types;
+
+pub async fn handle_event(
+    context: &AppContext,
+    event: ClickHouseEvent,
+    offset: Offset,
+) -> Result<Option<ClickHouseEvent>, UnhandledError> {
+    let mut props = match get_props(&event) {
+        Ok(r) => r,
+        Err(e) => {
+            offset.store().unwrap();
+            warn!("Failed to get props: {}", e);
+            // Drop exceptions with no properties - our users don't care about them, and neither do we.
+            return Ok(None);
+        }
+    };
+
+    let exceptions = match take_exception_list(event.uuid, &mut props) {
+        Ok(r) => r,
+        Err(e) => {
+            offset.store().unwrap();
+            warn!("Failed to take exception list: {}", e);
+            // Some exceptions don't have an exception list - if that's the case, we just pass them on.
+            // TODO - we should probably drop these exceptions to, I'm not sure
+            return Ok(Some(event));
+        }
+    };
+
+    let mut results = Vec::new();
+    for exception in exceptions.into_iter() {
+        // If we get an unhandled error during exception processing, we drop the offset without storing it,
+        // which means we'll take lag, but won't lose data.
+        results.push(process_exception(context, event.team_id, exception).await?);
+    }
+
+    props.fingerprint = Some(generate_fingerprint(&results));
+    props.exception_list = Some(results);
+
+    // "But we can open the box" said toad.
+    let mut event = event;
+    event.properties = Some(serde_json::to_string(&props).unwrap());
+
+    // We've processed the exception event succesffully, so we can store the offset.
+    offset.store().unwrap();
```

Comment:
Do we want to be storing the offset before we've written to kafka?

Reply:
Yes, I think so - once we have kafka emitting we'll probably do that here too, or maybe pass the final event and the offset along to some buffered flush